### PR TITLE
[DNM]refactor not leader

### PR DESCRIPTION
**Pull request by morefreeze**
PTAL

### CODE REVIEW

File: src/server/kv.rs [hunk 1]
```diff
@@ -306,13 +325,29 @@ impl StoreHandler {
                     item.set_res_type(res_type);
                     items.push(item);
                 }
+                cmd_prewrite_resp.set_results(RepeatedField::from_vec(items));
             }
             Err(e) => {
-                error!("storage error: {:?}", e);
+                if let StorageError::Engine(EngineError::Request(ref err)) = e {
```
Comment (ngaut) [inline src/server/kv.rs]: Extract to a function.

File: src/server/kv.rs [hunk 2]
```diff
@@ -208,19 +206,12 @@ impl StoreHandler {
...
-                        res_type.set_lock_info(lock_info);
-                    } else {
-                        let err_str = "key is locked but primary info not found".to_owned();
-                        error!("{}", err_str);
-                        res_type.set_field_type(ResultType_Type::Other);
-                        res_type.set_msg(err_str);
-                    }
+                } else if let Some((_, primary, ts)) = r.get_lock() {
+                    res_type.set_field_type(ResultType_Type::Locked);
+                    let mut lock_info = LockInfo::new();
```
Comment (ngaut) [inline src/server/kv.rs]: You should create a helper function to return a new LockInfo. DRY!

**Comment by ngaut** [comment]
PTAL @siddontang @disksing

File: src/server/kv.rs [hunk 3]
```diff
@@ -355,6 +321,9 @@ impl StoreHandler {
         let mut resp = Response::new();
         let mut cmd_commit_get_resp = CmdCommitThenGetResponse::new();
         cmd_commit_get_resp.set_ok(r.is_ok());
+        if let Err(ref e) = r {
```
Comment (siddontang) [inline src/server/kv.rs]: use match

File: src/storage/mvcc/txn.rs [hunk 4]
```diff
@@ -101,11 +101,13 @@ impl<'a, T: Engine + ?Sized> MvccTxn<'a, T> {
         }
         // ... or locks at any timestamp.
         if let Some(lock) = meta.get_lock() {
-            return Err(Error::KeyIsLocked {
-                key: key.raw().to_owned(),
-                primary: lock.get_primary_key().to_vec(),
-                ts: lock.get_start_ts(),
-            });
+            if lock.get_start_ts() != self.start_ts {
```
Comment (siddontang) [inline src/storage/mvcc/txn.rs]: <= ?

File: src/server/kv.rs (hunk 3 above)
Comment (morefreeze) [inline src/server/kv.rs]: fix in #378

File: src/storage/mvcc/txn.rs (hunk 4 above)
Comment (morefreeze) [inline src/storage/mvcc/txn.rs]: != is right @disksing

Comment (disksing) [inline src/storage/mvcc/txn.rs]: Yes. We want to make the prewrite operation idempotent. But I think we should return `Ok(())` here if timestamps are matched. @morefreeze

**Pull request merged**